devices/HumanDataCollector/HumanDataCollector.cpp

```diff
@@ -0,0 +1,663 @@
+        }
+
+        if (!pImpl->stateJointNamesDataPort.isClosed()) {
+            pImpl->stateJointNamesDataPort.close();
+        }
+
+        if (!pImpl->jointPositionsDataPort.isClosed()) {
+            pImpl->jointPositionsDataPort.close();
+        }
+
+        if (!pImpl->jointVelocitiesDataPort.isClosed()) {
+            pImpl->jointVelocitiesDataPort.close();
+        }
+
+        if (!pImpl->comPositionDataPort.isClosed()) {
+            pImpl->comPositionDataPort.close();
+        }
+
+        if (!pImpl->comVelocityDataPort.isClosed()) {
+            pImpl->comVelocityDataPort.close();
+        }
+
+        if (!pImpl->comProperAccelerationInBaseFrameDataPort.isClosed()) {
+            pImpl->comProperAccelerationInBaseFrameDataPort.close();
+        }
+
+        if (!pImpl->comProperAccelerationInWorldFrameDataPort.isClosed()) {
+            pImpl->comProperAccelerationInWorldFrameDataPort.close();
+        }
+
+        pImpl->iHumanState = nullptr;
+        pImpl->isStateProviderDevice = false;
+
+    }
+
+    if (pImpl->isWrenchProviderDevice) {
+
+        if (!pImpl->wrenchMeasurementSourceNamesDataPort.isClosed()) {
+            pImpl->wrenchMeasurementSourceNamesDataPort.close();
+        }
+
+        if (!pImpl->wrenchMeasurementValuesDataPort.isClosed()) {
+            pImpl->wrenchMeasurementValuesDataPort.close();
+        }
+
+        pImpl->iHumanWrenchMeasurements = nullptr;
+        pImpl->isWrenchProviderDevice = false;
+    }
+
+    if (pImpl->isDynamicsEstimatorDevice) {
+
+        if (!pImpl->wrenchEstimateSourceNamesDataPort.isClosed()) {
+            pImpl->wrenchEstimateSourceNamesDataPort.close();
+        }
+
+        if (!pImpl->wrenchEstimateValuesDataPort.isClosed()) {
+            pImpl->wrenchEstimateValuesDataPort.close();
+        }
+
+        if (!pImpl->dynamicsJointNamesDataPort.isClosed()) {
+            pImpl->dynamicsJointNamesDataPort.close();
+        }
+
+        if (!pImpl->jointTorquesDataPort.isClosed()) {
+            pImpl->jointTorquesDataPort.close();
+        }
+
+        pImpl->iHumanWrenchEstimates = nullptr;
+        pImpl->iHumanDynamics = nullptr;
+        pImpl->isDynamicsEstimatorDevice = false;
+    }
+
+    return true;
+}
+
+bool HumanDataCollector::attachAll(const yarp::dev::PolyDriverList &driverList)
+{
+    bool attachStatus = true;
+
+    if (driverList.size() > 3) {
+        yError() << LogPrefix << "This wrapper accepts at most three attached PolyDriver devices";
+        return false;
+    }
+
+    for (size_t i = 0; i < driverList.size(); i++) {
+        const yarp::dev::PolyDriverDescriptor* driver = driverList[i];
+
+        if (!driver) {
+            yError() << LogPrefix << "Passed PolyDriver device is nullptr";
+            return false;
+        }
+
+        attachStatus = attachStatus && attach(driver->poly);
+    }
+
+    // ====
+    // MISC
+    // ====
+
+    // Start the PeriodicThread loop
+    if (attachStatus && !start()) {
```

Comment:
Maybe:

```cpp
if (!(attachStatus && start()) {
```

In this way, if the attach status is false, the loop does not neither attempts to start since the device would fail anyway.

Reply:
>  the loop does not neither attempts to start since the device would fail anyway

I do not understand this. the variable `attachStatus` value (false/true) has not effect on `start()` call right?

Reply:
`if (attachStatus && !start()) {`

This means: if all the interfaces have been attached and there was a failure in starting the device, return false.

If there was an error attaching a device, the device is not started and the flow continues. In this specific case, there's nothing else after this if, and it's ok. But if in the future someone will add some code, it could be dangerous executing it because you already know that you already had a failure during the attach.

If you change the if as I suggested, the flow would instead stop at line 654.

Remember that if you have an `&&` in a condition, if the first statement is false, the second is not evaluated. This is because `a and b` is always false if a is false, for all values of b.